# 01. Importing libraries and data sets

Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

Creating a Path

In [2]:
path = r'C:\Users\chris\Desktop\Upskilling\CareerFoundry\Data Immersion\Achievement 4\08-2025 Instacart Basket Analysis'


Importing data sets

In [6]:
df_ords = pd.read_pickle(os.path.join(path, '02. Data', 'Prepared Data', 'order_products_merged.pkl'))

In [9]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order first_order  product_id  add_to_cart_order  \
0                     0.0         yes         196                  1   
1                     0.0         yes       14084                  2   
2                     0.0         yes       12427                  3   
3                     0.0         yes       26088                  4   
4                     0.0         yes       26405                  5   

   reordered                             product_name  department_id _merge  
0          0                                     Soda            7.0   both  
1          0  Organic Unsweetened Vanilla Almond Milk           16.0   both  
2          0                      Original Beef Jerky           19.0   both  
3          0               Aged White Cheddar Popcorn           19.0   both  
4          0         XL Pick-A-Size Paper Towel Rolls           17.0   both

In [11]:
df_dept = pd.read_csv(os.path.join(path, '02. Data', 'Prepared Data', 'departments_wrangled.csv'), index_col = False)

In [12]:
df_dept.head()

Unnamed: 0 department
0           1     frozen
1           2      other
2           3     bakery
3           4    produce
4           5    alcohol

In [18]:
df_dept.rename(columns = {'Unnamed: 0' : 'department_id'}, inplace = True)

In [19]:
df_dept.head()

department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [20]:
df_dept

department_id       department
0               1           frozen
1               2            other
2               3           bakery
3               4          produce
4               5          alcohol
5               6    international
6               7        beverages
7               8             pets
8               9  dry goods pasta
9              10             bulk
10             11    personal care
11             12     meat seafood
12             13           pantry
13             14        breakfast
14             15     canned goods
15             16       dairy eggs
16             17        household
17             18           babies
18             19           snacks
19             20             deli
20             21          missing

In [22]:
df_ords['meat_eater'] = df_ords['department_id'].apply(lambda x: 'Yes' if x == 12 else 'No')

In [27]:
df_ords['department_id'] == '12'

0           False
1           False
2           False
3           False
4           False
            ...  
32435054    False
32435055    False
32435056    False
32435057    False
32435058    False
Name: department_id, Length: 32435059, dtype: bool

creating a subset of the df ie only 1000000 rows

In [28]:
df = df_ords[:1000000]

In [29]:
df_ords_checked = pd.read_csv(os.path.join(path, '02. Data', 'Prepared Data', 'order_checked.csv'))

In [30]:
df_ords_checked.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id eval_set  order_number  \
0             0           0   2539329        1    prior             1   
1             1           1   2398795        1    prior             2   
2             2           2    473747        1    prior             3   
3             3           3   2254736        1    prior             4   
4             4           4    431534        1    prior             5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order first_order  
0                   2                  8                     0.0         yes  
1                   3                  7                    15.0          no  
2                   3                 12                    21.0          no  
3                   4                  7                    29.0          no  
4                   4                 15                    28.0          no

In [31]:
df_prods_checked = pd.read_csv(os.path.join(path, '02. Data', 'Prepared Data', 'products_checked.csv'))

In [32]:
df_prods_checked.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [37]:
df_merged_ords_prices = df_ords.merge(df_prods_checked[["product_id", "prices"]], on = 'product_id', how = "left", indicator = '_merge_prices')

In [38]:
df_merged_ords_prices.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order first_order  product_id  add_to_cart_order  \
0                     0.0         yes         196                  1   
1                     0.0         yes       14084                  2   
2                     0.0         yes       12427                  3   
3                     0.0         yes       26088                  4   
4                     0.0         yes       26405                  5   

   reordered                             product_name  department_id _merge  \
0          0                                     Soda            7.0   both   
1          0  Organic Unsweetened Vanilla Almond Milk           16.0   both   
2          0                      Original Beef Jerky           19.0   both   
3          0               Aged White Cheddar Popcorn           19.0   both   
4          0         XL Pick-A-Size Paper Towel Rolls           17.0   both   

  meat_eater  prices _merge_prices  
0         No     9.0          both  
1         No    12.5          both  
2         No     4.4          both  
3         No     4.7          both  
4         No     1.0          both

In [39]:
df = df_merged_ords_prices[:1000000]

Creating a user-defined function for price

In [40]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [42]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\chris\AppData\Local\Temp\ipykernel_15860\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [45]:
df.value_counts('price_range')

price_range
Mid-range product    672523
Low-range product    314111
High range            12413
Not enough data         953
Name: count, dtype: int64

Checking the highest value

In [46]:
df['prices'].max()

99999.0

In [50]:
print(df[df['prices'] >= 100])

        order_id  user_id  order_number  orders_day_of_week  \
1576      912404       17            12                   2   
1638      603376       17            22                   6   
16534    3264360      135             2                   2   
16540     892534      135             3                   0   
53714     229704      342             8                   1   
...          ...      ...           ...                 ...   
959909    840159     6217            13                   1   
959955   1235750     6217            20                   5   
961209     40712     6221             5                   0   
961283   2567369     6221            10                   2   
961444    842052     6222            10                   1   

        order_hour_of_day  days_since_prior_order first_order  product_id  \
1576                   14                     5.0          no       21553   
1638                   16                     4.0          no       21553   
16534       

At this stage i have noticed that there appears to be one product with a price of 14900, which must be wrong... i assume that it should be 1.49. Perhaps i will need to correct this later. There is also a product with a price of 99999 which must also be wrong, but it might be that the price is missing and therefore this value has been imputed. this might require further investigation.

Deriving a new variable using loc()

In [51]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\chris\AppData\Local\Temp\ipykernel_15860\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [52]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [54]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [55]:
df.head

<bound method NDFrame.head of         order_id  user_id  order_number  orders_day_of_week  \
0        2539329        1             1                   2   
1        2539329        1             1                   2   
2        2539329        1             1                   2   
3        2539329        1             1                   2   
4        2539329        1             1                   2   
...          ...      ...           ...                 ...   
999995   1876509     6467            51                   2   
999996   1876509     6467            51                   2   
999997   1876509     6467            51                   2   
999998   1876509     6467            51                   2   
999999   1876509     6467            51                   2   

        order_hour_of_day  days_since_prior_order first_order  product_id  \
0                       8                     0.0         yes         196   
1                       8                     0.0         y

In [57]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     672523
Low-range product     314111
High-range product     12413
NaN                      953
Name: count, dtype: int64

performing the loc() function on the whole df

In [58]:
df_merged_ords_prices.loc[df_merged_ords_prices['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [59]:
df_merged_ords_prices.loc[(df_merged_ords_prices['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [60]:
df_merged_ords_prices.loc[df_merged_ords_prices['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [61]:
df_merged_ords_prices['price_range_loc'].value_counts(dropna = False)

price_range_loc
NaN                   21219115
Low-range product     10126883
Mid-range product       672523
High-range product      417678
Name: count, dtype: int64

printing the frequency of a column

In [62]:
df_merged_ords_prices['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6210092
1    5666141
6    4500565
2    4218008
5    4209743
3    3844291
4    3787359
Name: count, dtype: int64

creating an empty list

In [66]:
result = []
for value in df_merged_ords_prices["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [67]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

combining this list with the df by creating a new variable

In [68]:
df_merged_ords_prices['busiest_day'] = result

In [69]:
df_merged_ords_prices['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22438748
Busiest day        6210092
Least busy         3787359
Name: count, dtype: int64

Creating a new empty list for days

In [73]:
result2 = []
for value in df_merged_ords_prices["orders_day_of_week"]:
  if value in [0, 1]:
    result2.append("Busiest days")
  elif value in [3, 4]:
    result2.append("Slowest days")
  else:
    result2.append("Regularly busy")

In [74]:
result2

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest day

Merging the list to the df

In [76]:
df_merged_ords_prices['busiest_days'] = result2

Printing the frequency for the new column

In [77]:
df_merged_ords_prices['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12928316
Busiest days      11876233
Slowest days       7631650
Name: count, dtype: int64

deleting the unnecessary column busiest_day

In [78]:
df_merged_ords_prices.drop(columns=['busiest_day'], inplace=True)

Identifying the hours with the most orders

In [80]:
df_merged_ords_prices['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2764576
11    2738777
14    2691698
15    2664683
13    2663454
12    2621000
16    2537602
9     2456827
17    2089600
8     1720027
18    1638022
19    1259446
20     977071
7      891979
21     796397
22     634761
23     402623
6      290798
0      218957
1      115789
5       88068
2       69437
4       53286
3       51321
Name: count, dtype: int64

creating an empty list

In [81]:
result3 = []
for value in df_merged_ords_prices["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result3.append("Most orders")
  elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
    result3.append("Average order")
  else:
    result3.append("Fewest orders")

In [82]:
result3

['Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order',
 'Average order'

Merging the list to the df

In [83]:
df_merged_ords_prices['busiest_period_of_day'] = result3

Printing the frequency for the new column

In [85]:
df_merged_ords_prices['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders      21138617
Average order    10007303
Fewest orders     1290279
Name: count, dtype: int64

In [86]:
df_merged_ords_prices.to_pickle(os.path.join(path, '02. Data','Prepared Data', 'order_products_new_variables.pkl'))